text/XXXX-externalize_private_keys.md

```diff
@@ -0,0 +1,111 @@
+[reference-level-explanation]: #reference-level-explanation
+
+Doing the work of generalizing secret storage can be handled by an external package: https://github.com/zalando/go-keyring. go-keyring supports Windos, OS X, and Linux. The key is stored in a `(username,password)` combination, with the "username" being the user's ProfileID (hash of public key), and the password being the base64-encoded private key. OS Keyrings support namespacing, qri's namespace will be a reversed-domain-name constant:
+
+```go
+const qriKeyringServiceID = "io.qri.keys"
+```
+
+Once setup is complete, `config.Profile.ID` stores the ProfileID string this configuration is attached to. Starting a qri process uses this as the "username" in a `Get` call:
+
+```go
+privKey, _ := keyring.Get(qriKeyringServiceID, config.Profile.ID)
+```
+
+### Keys subsystem
+
+All of this work should happen in a new package within the qri codebase: `github.com/qri-io/qri/keys`. We need to takle key rotation next, followed by storing _many_ keys for doing access control. All of this work should happen in a new subsystem called `keys.Service`. `lib.Instance` gets a new field that points to this service.
+
+### Phasing out `config.Profile.PrivKey`
+This work should happen in phases:
+
+1. **copy to keyring, support env var**. In in the next release, add in the new APIs, prefer their use when present.
+2. **intiailize with an empty privKey value**. Once we know the keyring approach works across a number of different platforms in a wide release, switch default initialization behaviour to _not_ write the privKey value.
+3. **remove the field** Eventually, remove the field entirely with a repo migration.
+
+In all phases the privKey field should be the _lowest_ in the loading hierarchy. Keyring values will override `Profile.PrivKey`, and the env var will override all other values.
+
+### Recovering from a deleted `.qri` folder
+Once we've externalized the private keys, we can change the behaviour of `qri setup` to first scan the keyring under the `io.qri.keys` namespace for existing keys. If a key is found, we can ask the user if they'd like to create a new account, or recover an old one.
+
+We can bolster this with calls to the default registry for the discovered public keys to add username info to the list. Profile information may not exist on the default registry, so we'll need to default to showing just the raw ProfileID. We can support this process with docs.
+
+Instead of jumping straight to building the UI for switching, we can supply a `--recover` flag that lets a user supply a ProfileID to recover from:
+
+```
+$ qri setup --recover QmProfileID
+```
+
+listing available private keys in a keyring may also form the basis of profile switching in the future.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+### More OS-Specific Work.
+This increases our reliance on an OS-level API that sometimes doesn't exist, and runs the risk of deviating behaviour between platforms. 
```

Comment:
I have been a _heavy_ macOS user over the years. I almost never use Keychain. If I used Keychain Access today, I’d have to poke around to find what I needed.

I know some novices who use it heavily (maybe because of Safari integration and sync), and I know super power users who wouldn’t know it any better than I do because they stick everything in 1Password or such.

Neither is key storage, I’m just talking about usage/familiarity.

OS native is good. But unless you need the user to look at the keys and interact with them, I don’t know if it gains anything from UX.

And that’s only one platform of the three.

I think Qri needs to do most of the key UX itself. No?

In which case, platform-native wouldn’t gain much from UX.

Platform-native could turn out to be easiest/best way anyway.

I don’t know anything of storage in `.ssh` but good to know there is something standardish.

Reply:
I should clarify this in the RFC. Storing in OS keychain managers is only solving the problem of _"a safe place outside qri to store keys"_. The benefit of using this system is leveraging OS level access control & encryption at rest. Totally agreed we have to build our own UX for key management. If anything we _lose_ a little on the UX side, as users can delete their keys from a place outside Qri, which may cause problems.

> I don’t know anything of storage in .ssh but good to know there is something standardish.
`.ssh` isn't a standard, it's just an application with a lot of use. We've run into problems piggybacking off other software without using the binary itself for interaction in the past, so I'd like to avoid any solution that stores keys in a place intended for another program